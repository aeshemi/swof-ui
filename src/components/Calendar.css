.rotation {
    border-radius: 0;
    border: none;
}

.morning {
    background-color: #6593f5;
}

.afternoon {
    background-color: #dc6751;
}
