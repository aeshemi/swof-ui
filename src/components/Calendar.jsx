import React, { Component } from 'react';
import BigCalendar from 'react-big-calendar';
import moment from 'moment';
import dates from 'react-big-calendar/lib/utils/dates';
import ApiClient from '../ApiClient';
import 'react-big-calendar/lib/css/react-big-calendar.css';
import './Calendar.css';

BigCalendar.setLocalizer(BigCalendar.momentLocalizer(moment));

const setVisibleDates = date => ({
  firstVisible: moment(dates.firstVisibleDay(date)),
  lastVisible: moment(dates.lastVisibleDay(date))
});

class Calendar extends Component {
  constructor(props) {
    super(props);
    this.state = { ...setVisibleDates(moment()), schedule: [] };
    this.getSchedule = this.getSchedule.bind(this);
    this.setCalendarDates = this.setCalendarDates.bind(this);
  }

  componentDidMount() {
    this.getSchedule();
  }

  getSchedule() {
    ApiClient.getCalendar(this.state.firstVisible.format('YYYY-MM-DD')).then(data => {
      const schedule = data.map(s => ({
        ...s,
        startDateTime: new Date(s.startDateTime),
        endDateTime: new Date(s.endDateTime)
      }));

      this.setState({ schedule });
    });
  }

  setCalendarDates(date) {
    if (date === undefined) return;

    const { firstVisible, lastVisible } = this.state;

    const newDates = setVisibleDates(date);

    if (
      firstVisible.isSame(newDates.firstVisible, 'day') ||
      lastVisible.isSame(newDates.lastVisible, 'day')
    )
      return;

    this.setState({ ...newDates }, this.getSchedule);
  }

  render() {
    return (
      <BigCalendar
        defaultView="month"
        events={this.state.schedule}
        startAccessor="startDateTime"
        endAccessor="endDateTime"
        titleAccessor="name"
        style={{ height: '100vh' }}
        views={['month', 'week']}
        eventPropGetter={event => ({ className: `rotation ${event.shift.toLowerCase()}` })}
        onNavigate={this.setCalendarDates}
      />
    );
  }
}

export default Calendar;
