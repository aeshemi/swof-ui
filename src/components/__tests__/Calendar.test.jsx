import Calendar from '../Calendar';

describe('Calendar', () => {
  beforeAll(() => {
    fetch.mockResponse(JSON.stringify([]));
  });

  afterAll(() => {
    fetch.resetMocks();
  });

  it('should call getSchedule during componentDidMount', () => {
    const spy = jest.spyOn(Calendar.prototype, 'getSchedule');

    mount(<Calendar />);

    expect(spy).toHaveBeenCalled();
    spy.mockReset();
    spy.mockRestore();
  });

  it('BigCalendar.eventPropGetter returns correct style', () => {
    const wrapper = shallow(<Calendar />);

    expect(
      wrapper
        .find('Uncontrolled(Calendar)')
        .props()
        .eventPropGetter({ shift: 'Morning' }).className
    ).toBe('rotation morning');
  });

  it('BigCalendar.onNavigate calls setCalendarDates', () => {
    const spy = jest.spyOn(Calendar.prototype, 'setCalendarDates');
    const wrapper = shallow(<Calendar />);

    wrapper
      .find('Uncontrolled(Calendar)')
      .props()
      .onNavigate();

    expect(spy).toHaveBeenCalled();

    spy.mockReset();
    spy.mockRestore();
  });

  it('setCalendarDates only calls getSchedule when visible dates changed', () => {
    const spy = jest.spyOn(Calendar.prototype, 'getSchedule');
    const instance = shallow(<Calendar />).instance();

    const { firstVisible, lastVisible } = instance.state;

    instance.setCalendarDates(undefined);
    instance.setCalendarDates(firstVisible.add(5, 'd').toDate());
    instance.setCalendarDates(lastVisible.subtract(5, 'd').toDate());
    instance.setCalendarDates(firstVisible.add(2, 'M').toDate());
    instance.setCalendarDates(lastVisible.add(2, 'M').toDate());

    expect(spy).toHaveBeenCalledTimes(3); // 2 + componentDidMount call

    spy.mockReset();
    spy.mockRestore();
  });

  it('renders correctly with no schedules', () => {
    const wrapper = shallow(<Calendar />);
    expect(wrapper).toMatchSnapshot();
  });

  it('renders correctly with schedules', () => {
    const wrapper = shallow(<Calendar />);

    wrapper.setState({
      schedule: [
        {
          startDateTime: '2018-08-20T09:00:00',
          endDateTime: '2018-08-20T13:00:00',
          name: 'Hetti Whatley',
          shift: 'Morning'
        },
        {
          startDateTime: '2018-08-20T13:00:00',
          endDateTime: '2018-08-20T17:00:00',
          name: 'Natalee Wigin',
          shift: 'Afternoon'
        },
        {
          startDateTime: '2018-08-21T09:00:00',
          endDateTime: '2018-08-21T13:00:00',
          name: 'Charlot Eede',
          shift: 'Morning'
        },
        {
          startDateTime: '2018-08-21T13:00:00',
          endDateTime: '2018-08-21T17:00:00',
          name: 'Monti Vercruysse',
          shift: 'Afternoon'
        },
        {
          startDateTime: '2018-08-22T09:00:00',
          endDateTime: '2018-08-22T13:00:00',
          name: 'Nicolea Appleyard',
          shift: 'Morning'
        },
        {
          startDateTime: '2018-08-22T13:00:00',
          endDateTime: '2018-08-22T17:00:00',
          name: 'Nap Nimmo',
          shift: 'Afternoon'
        },
        {
          startDateTime: '2018-08-23T09:00:00',
          endDateTime: '2018-08-23T13:00:00',
          name: 'Lethia Bonefant',
          shift: 'Morning'
        },
        {
          startDateTime: '2018-08-23T13:00:00',
          endDateTime: '2018-08-23T17:00:00',
          name: 'Sharity Dronsfield',
          shift: 'Afternoon'
        }
      ]
    });

    wrapper.update();
    expect(wrapper).toMatchSnapshot();
  });
});
