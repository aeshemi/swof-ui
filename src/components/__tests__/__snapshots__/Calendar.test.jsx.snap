// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Calendar renders correctly with no schedules 1`] = `
<Uncontrolled(Calendar)
  defaultView="month"
  endAccessor="endDateTime"
  eventPropGetter={[Function]}
  events={Array []}
  onNavigate={[Function]}
  startAccessor="startDateTime"
  style={
    Object {
      "height": "100vh",
    }
  }
  titleAccessor="name"
  views={
    Array [
      "month",
      "week",
    ]
  }
/>
`;

exports[`Calendar renders correctly with schedules 1`] = `
<Uncontrolled(Calendar)
  defaultView="month"
  endAccessor="endDateTime"
  eventPropGetter={[Function]}
  events={
    Array [
      Object {
        "endDateTime": "2018-08-20T13:00:00",
        "name": "Hetti Whatley",
        "shift": "Morning",
        "startDateTime": "2018-08-20T09:00:00",
      },
      Object {
        "endDateTime": "2018-08-20T17:00:00",
        "name": "Natalee Wigin",
        "shift": "Afternoon",
        "startDateTime": "2018-08-20T13:00:00",
      },
      Object {
        "endDateTime": "2018-08-21T13:00:00",
        "name": "Charlot Eede",
        "shift": "Morning",
        "startDateTime": "2018-08-21T09:00:00",
      },
      Object {
        "endDateTime": "2018-08-21T17:00:00",
        "name": "Monti Vercruysse",
        "shift": "Afternoon",
        "startDateTime": "2018-08-21T13:00:00",
      },
      Object {
        "endDateTime": "2018-08-22T13:00:00",
        "name": "Nicolea Appleyard",
        "shift": "Morning",
        "startDateTime": "2018-08-22T09:00:00",
      },
      Object {
        "endDateTime": "2018-08-22T17:00:00",
        "name": "Nap Nimmo",
        "shift": "Afternoon",
        "startDateTime": "2018-08-22T13:00:00",
      },
      Object {
        "endDateTime": "2018-08-23T13:00:00",
        "name": "Lethia Bonefant",
        "shift": "Morning",
        "startDateTime": "2018-08-23T09:00:00",
      },
      Object {
        "endDateTime": "2018-08-23T17:00:00",
        "name": "Sharity Dronsfield",
        "shift": "Afternoon",
        "startDateTime": "2018-08-23T13:00:00",
      },
    ]
  }
  onNavigate={[Function]}
  startAccessor="startDateTime"
  style={
    Object {
      "height": "100vh",
    }
  }
  titleAccessor="name"
  views={
    Array [
      "month",
      "week",
    ]
  }
/>
`;
